api/layer/object_cache.go

```diff
@@ -1,79 +1,96 @@
 package layer
 
 import (
+	"context"
 	"sync"
 	"time"
 
 	cid "github.com/nspcc-dev/neofs-api-go/pkg/container/id"
 )
 
 /*
-	This is an implementation of a cache for ListObjectsV2 which we return to users by ContinuationToken.
+	This is an implementation of a cache for ListObjectsV2/V1 which we return to users by ContinuationToken/Marker.
 
 	The cache is a map which has a key: (access_key from AccessBox) + container_id and a value: list of objects with
 	creation time. After putting a record we start a timer (via time.AfterFunc) that removes the record after
 	defaultCacheLifetime value.
 
-	ContinuationToken in our gateway is an objectID in NeoFS.
+	ContinuationToken in our gateway is an objectID in NeoFS. Marker is a key of an object.
 
-	We don't keep ContinuationToken in this structure because we assume that users who received the token can reconnect
-	to other gateways and they should be able to get a list of objects.
+	We don't keep ContinuationToken/Marker in this structure because we assume that users who received the token can
+	reconnect to other gateways and they should be able to get a list of objects.
 	When we receive the token from the user we just try to find the cache and then we return the list of objects which
-	starts from this token (i.e. objectID).
+	starts after this token (i.e. objectID or key).
 */
 
-// ObjectsListV2Cache provides interface for cache of ListObjectsV2 in a layer struct.
+// ObjectsListCache provides interface for cache of ListObjectsV2 in a layer struct.
 type (
-	ObjectsListV2Cache interface {
-		Get(token string, key string) []*ObjectInfo
-		Put(key string, objects []*ObjectInfo)
+	ObjectsListCache interface {
+		Get(key cacheOptions) []*ObjectInfo
+		Put(p cacheOptions, objects []*ObjectInfo)
 	}
 )
 
-var (
-	defaultCacheLifetime = time.Second * 60
-)
+const defaultCacheLifetime = time.Second * 60
 
 type (
 	listObjectsCache struct {
-		caches map[string]cache
-		mtx    sync.RWMutex
+		cacheLifetime time.Duration
+		caches        map[cacheOptions]cache
+		mtx           sync.RWMutex
 	}
 	cache struct {
 		list []*ObjectInfo
 	}
+	cacheOptions struct {
+		key       string
+		delimiter string
+		prefix    string
+	}
 )
 
-func newListObjectsCache() *listObjectsCache {
+func newListObjectsCache(lifetime time.Duration) *listObjectsCache {
 	return &listObjectsCache{
-		caches: make(map[string]cache),
+		caches:        make(map[cacheOptions]cache),
+		cacheLifetime: lifetime,
 	}
 }
 
-func (l *listObjectsCache) Get(token, key string) []*ObjectInfo {
+func (l *listObjectsCache) Get(key cacheOptions) []*ObjectInfo {
 	l.mtx.RLock()
 	defer l.mtx.RUnlock()
 	if val, ok := l.caches[key]; ok {
-		return trimAfterObjectID(token, val.list)
+		return val.list
```

Comment:
Why we refused `trimAfterObjectID`?

Reply:
At first, earlier cache was connected closely with token/marker.  In the previous version we read cache only if we received token/marker from the request. 
Now we read cache always,  both versions can use the same cache entry with one big unfiltered list of objects.
That's why specific functions for each version become more senseless.

At second, in the previous version we did the following actions:
1. Try to get cache: find the entry, filter by token/marker
2. If we can't find the entry we ask NeoFS for the list and filter by token/marker
We made filtering in both cases.

At third, one common function for both versions allowed us to create a separate function `listAllObjects` and use it in both ListObjects, what make these functions shorter.

That's why I decided to  take filtering by token/marker outside and replace `GetByContinuationToken` and `GetByMarker` by `Get` which returns all of objects.